File: backend/src/services/chunkingservice.js
import { chunkText } from "../utils/chunker.js";
import { ChunkRepository } from "../repositories/chunkrepo.js";
import { indexChunks } from "./indexingservice.js";

const repo = new ChunkRepository();

export async function chunkDocument({ document, userId }) {
  // 1. Delete old chunks (safe re-run)
  await repo.deleteByDocument({
    documentId: document.id,
    userId
  });

  // 2. Create new chunks (LangChain)
  const chunks = await chunkText(document.content);
  console.log("Chunks:", chunks.length);
console.log("First chunk content:", JSON.stringify(chunks[0]?.content));
console.log("Content length:", chunks[0]?.content?.length);

  // 3. Store chunks in Postgres
  await repo.bulkCreate({
    documentId: document.id,
    userId,
    chunks
  });

  // 4. Fetch stored chunks (with IDs)
  const storedChunks = await repo.findByDocument({
    documentId: document.id,
    userId
  });

  // 5. Index chunks into Qdrant
  await indexChunks({ chunks: storedChunks });

  return { count: storedChunks.length };
}

File: backend/src/services/chunking_service.py
from sqlalchemy.orm import Session

from src.db.models import Document
from src.repositories.chunk import ChunkRepository
from src.utils.chunker import chunk_text
from .indexing_service import index_chunks
from .vector_service import delete_document_vectors


repo = ChunkRepository()


def chunk_document(db: Session, *, document: Document, user_id: str) -> dict:
    delete_document_vectors(document_id=document.id, user_id=user_id)
    repo.delete_by_document(db, document_id=document.id, user_id=user_id)

    chunks = chunk_text(document.content)
    print("Chunks:", len(chunks))
    print("First chunk content:", chunks[0]["content"] if chunks else None)
    print("Content length:", len(chunks[0]["content"]) if chunks else None)

    repo.bulk_create(db, document_id=document.id, user_id=user_id, chunks=chunks)
    stored_chunks = repo.find_by_document(db, document_id=document.id, user_id=user_id)
    index_chunks(chunks=stored_chunks)

    return {"count": len(stored_chunks)}
